### PR TITLE
Allow collection and cursor to be safely contained in the same struct

### PR DESCRIPTION
**Pull request by sockmaster27**
At the moment it doesn't appear to be possible to create a struct which represents an owned collection and a position in it without the use of raw pointers. Ideally this would be done like this:
```Rust
struct Object {
    collection: RBTree<...>,
    cursor: Cursor<'?, ...>,
}
```
However, this isn't possible because of the self-referential nature of this struct.

Instead I propose a new type of cursor, `CursorOwning`, to coexist with `Cursor` and `CursorMut`. It would be mostly functionally equivalent to `CursorMut`, except its lifetime wouldn't be bounded by that of the collection, as it would own it rather than borrow it.

For now I've implemented this for the `RBTree` collection, and I'm willing to do the same for the rest of the library if accepted.

### CODE REVIEW

**Comment by Amanieu** [comment]
To avoid massive code duplication, have you considered using a crate like [ouroboros](https://docs.rs/ouroboros/) instead?

**Comment by sockmaster27** [comment]
Yes, that does make it technically possible, however, it adds a lot of complexity. First of all, if any control of the interface exposed by the struct is desired, a new struct would have to be introduced that would basically just be a reimplementation of `CursorOwning`:
```Rust
#[self_referencing]
struct CursorOwning<A: Adapter + 'static>
where
    <A as intrusive_collections::Adapter>::LinkOps: RBTreeOps,
{
    tree: RBTree<A>,
    #[borrows(mut tree)]
    #[covariant]
    cursor: CursorMut<'this, A>,
}
```
Removing the excess generics would of course cut this down slightly. 
This struct has the interface generated by Ouroboros instead of a custom one, which results in some slight oddities, like how accessing the `&mut CursorOwning` can only be done in a closure through `with_cursor_mut()`, and construction is done like:
```Rust
CursorOwning::new(RBTree::new(MyAdapter::new()), |tree| {
    tree.cursor_mut()
}
```
A point that I've only thought about just now is that things only get worse if `Send` and `Sync` are involved, as they are in my particular case. I don't see any reason for why a `CursorOwning` shouldn't be at least `Send` if the tree is as well (in contrast to `CursorMut`). Here I don't any way around an unsafe implementation of `Send`:
```Rust
unsafe impl<A: Adapter + Send> Send for CursorOwning<A>
where
    <A::PointerOps as PointerOps>::Pointer: Send,
    A::LinkOps: RBTreeOps,
{
}
```

Now if this is only relevant to me this is obviously fine, but I feel like this isn't a super obscure problem to have, and it would be a great thing for the library to offer.

**Comment by Amanieu** [comment]
My main concern with your proposed implementation of `CursorOwning` is the bloat caused by duplicating all of the methods of `CursorMut`. Instead would it be possible to have a different design when a `CursorOwning` doesn't provide these methods but instead provides a way to get a temporary `CursorMut`?

```rust
impl CursorOwning<A> {
    pub fn with_cursor<T>(&mut self, f: impl FnOnce(&mut CursorMut<A>) -> T) -> T {
        let mut cursor = self.tree.cursor_mut_from_ptr(self.ptr);
        let ret = f(&mut cursor);
        self.ptr = cursor.get();
        ret
    }
}
```

**Comment by sockmaster27** [comment]
Sure! This was an attempt to make `CursorMut` as similar as possible to the others, but it can certainly be done like that too.

File: src/rbtree.rs [hunk 1]
```diff
@@ -1927,10 +2063,56 @@ where
...
+    /// based on its key.
+    ///
+    /// Consumes the `RBTree` and returns an owning cursor pointing to the newly added element.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the new element is already linked to a different intrusive
+    /// collection.
+    #[inline]
+    pub fn insert_owning<'a>(
```
Comment (Amanieu) [inline src/rbtree.rs:2105]: I don't think this method is sufficiently useful to justify its inclusion. I would prefer not having it unless you have a specific use case for it.

File: src/rbtree.rs [hunk 2]
```diff
@@ -1449,6 +1448,61 @@ where
...
+    tree: RBTree<A>,
+}
+
+impl<A: Adapter> CursorOwning<A>
+where
+    A::LinkOps: RBTreeOps,
+{
+    /// Consumes self and returns the inner `RBTree`.
+    #[inline]
+    pub fn take_tree(self) -> RBTree<A> {
```
Comment (Amanieu) [inline src/rbtree.rs:1466]:
```suggestion
    pub fn into_inner(self) -> RBTree<A> {
```

**Review by Amanieu — COMMENTED** [review]
Could you add this to the linked list types as well for consistency?

File: src/rbtree.rs (hunk 1 above)
Comment (sockmaster27) [reply src/rbtree.rs:2105]: Well given that a `CursorMut` cannot be made into a `CursorOwning` it's pretty convenient, but not strictly necessary I guess?

File: src/linked_list.rs [hunk 3]
```diff
@@ -1064,6 +1063,61 @@ where
     }
 }
 
+/// A cursor with ownership over the `XorLinkedList` it points into.
```
Comment (Amanieu) [inline src/linked_list.rs:1066]:
```suggestion
/// A cursor with ownership over the `LinkedList` it points into.
```

File: src/singly_linked_list.rs [hunk 4]
```diff
@@ -840,6 +840,61 @@ where
     }
 }
 
+/// A cursor with ownership over the `XorLinkedList` it points into.
```
Comment (Amanieu) [inline src/singly_linked_list.rs:843]:
```suggestion
/// A cursor with ownership over the `SinglyLinkedList` it points into.
```

**Review by Amanieu — COMMENTED** [review]
LGTM! Just a few typos.

File: src/linked_list.rs [hunk 5]
```diff
@@ -16,7 +16,6 @@ use core::sync::atomic::{AtomicPtr, Ordering};
 use crate::link_ops::{self, DefaultLinkOps};
 use crate::pointer_ops::PointerOps;
 use crate::singly_linked_list::SinglyLinkedListOps;
-use crate::unchecked_option::UncheckedOptionExt;
```
Comment (Amanieu) [inline src/linked_list.rs:19]: You need to keep this so the code works on Rust 1.56.

**Comment by Amanieu** [comment]
CI is failing.

**Comment by Amanieu** [comment]
Thanks!

**Pull request merged**